src/weight_windows.cpp

```diff
@@ -624,20 +625,45 @@ void WeightWindows::update_magic(
   auto mesh_vols = this->mesh()->volumes();
 
   int e_bins = new_bounds.shape()[0];
-  for (int e = 0; e < e_bins; e++) {
-    // select all
-    auto group_view = xt::view(new_bounds, e);
 
-    // divide by volume of mesh elements
-    for (int i = 0; i < group_view.size(); i++) {
-      group_view[i] /= mesh_vols[i];
+  if (settings::solver_type == SolverType::MONTE_CARLO ||
+      !FlatSourceDomain::adjoint_) {
+    // If we are computing weight windows with forward fluxes derived from a
+    // Monte Carlo or forward random ray solve, we use the MAGIC algorithm.
+    for (int e = 0; e < e_bins; e++) {
+      // select all
+      auto group_view = xt::view(new_bounds, e);
+
+      // divide by volume of mesh elements
+      for (int i = 0; i < group_view.size(); i++) {
+        group_view[i] /= mesh_vols[i];
+      }
+
+      double group_max =
+        *std::max_element(group_view.begin(), group_view.end());
+      // normalize values in this energy group by the maximum value for this
+      // group
+      if (group_max > 0.0)
+        group_view /= 2.0 * group_max;
     }
+  } else {
+    // If we are computing weight windows with adjoint fluxes derived from an
+    // adjoint random ray solve, we use the FW-CADIS algorithm.
+    for (int e = 0; e < e_bins; e++) {
+      // select all
+      auto group_view = xt::view(new_bounds, e);
+
+      // divide by volume of mesh elements
+      for (int i = 0; i < group_view.size(); i++) {
+        group_view[i] /= mesh_vols[i];
+      }
+    }
```

Comment:
This block is exactly the same as on the top of the `if`. Any way we could join those together so as not to duplicate it?

Reply:
It's tough as the magic method has extra logic that gets applied within the loop over `e_bins`, whereas FW-CADIS has extra logic that gets applied outside of the loop for normalization. So, it seems like you'd have to have two lower level `if` statements instead of one higher level one if we wanted to avoid the repeated loops. Hard to say which is cleaner. I tend to lean towards keeping it how it is so it's easy to see what each method is doing without having to parse the inner if statements, but could be easily swayed!